subiquity/server/apt.py

```diff
@@ -50,58 +110,45 @@ async def mount(self, device, mountpoint, options=None, type=None):
             opts.extend(['-t', type])
         await self.app.command_runner.run(
             ['mount'] + opts + [device, mountpoint])
-        self._mounts.append(mountpoint)
+        m = Mountpoint(mountpoint=mountpoint)
+        self._mounts.append(m)
+        return m
 
     async def unmount(self, mountpoint):
         await self.app.command_runner.run(['umount', mountpoint])
 
-    async def setup_overlay(self, source, target):
+    async def setup_overlay(self, lowers):
         tdir = self.tdir()
-        w = f'{tdir}/work'
-        u = f'{tdir}/upper'
-        for d in w, u:
+        target = f'{tdir}/mount'
+        workdir = f'{tdir}/work'
+        upperdir = f'{tdir}/upper'
+        for d in target, workdir, upperdir:
             os.mkdir(d)
-        await self.mount(
-            'overlay', target, type='overlay',
-            options=f'lowerdir={source},upperdir={u},workdir={w}')
-        return u
-
-    async def configure(self, context):
-        # Configure apt so that installs from the pool on the cdrom are
-        # preferred during installation but not in the installed system.
-        #
-        # First we create an overlay ('configured') over the installation
-        # source and configure that overlay as we want the target system to
-        # end up by running curtin's apt-config subcommand.
-        #
-        # Then we create a fresh overlay ('for_install') over the first one
-        # and configure it for the installation. This means:
-        #
-        # 1. Bind-mounting /cdrom into this new overlay.
-        #
-        # 2. When the network is expected to be working, copying the original
-        #    /etc/apt/sources.list to /etc/apt/sources.list.d/original.list.
-        #
-        # 3. writing "deb file:///cdrom $(lsb_release -sc) main restricted"
-        #    to /etc/apt/sources.list.
-        #
-        # 4. running "apt-get update" in the new overlay.
-        #
-        # When the install is done we try to make the installed system's apt
-        # state look as if the pool had never been configured. So this means:
-        #
-        # 1. Removing /cdrom from the installed system.
-        #
-        # 2. Copying /etc/apt from the 'configured' overlay to the installed
-        #    system.
-        #
-        # 3. If the network is working, run apt-get update in the installed
-        #    system, or if it is not, just copy /var/lib/apt/lists from the
-        #    'configured' overlay.
-
-        self.configured = self.tdir()
-
-        config_upper = await self.setup_overlay(self.source, self.configured)
+
+        def lowerdir_for(lower):
+            if isinstance(lower, str):
+                return lower
+            if isinstance(lower, Mountpoint):
```

Comment:
We could simplify this sequence a little by adding a method that the `_MountBase`s implement

Reply:
I'm not sure quite what you meant here, but I refactored this a little anyway and in the process noticed a bug, so thanks for the suggestion :)